### PR TITLE
GL-670 Generate BPM.CSV from BPM file

### PR DESCRIPTION
**Pull request by gbggrant**
Added functionality to parse Illumina BPM (Bead Pool Manifest) files
Added new CLP to generate Illumina 'normalization manifest' CSV file'
Modified GtcToVcf and CompareGtcFiles to use IlluminaBPMFile parser (rather than using the old csv)

### Description

_Give your PR a **concise** yet **descriptive** title_
_Please explain the changes you made here._
_Explain the **motivation** for making this change. What existing problem does the pull request solve?_
_Mention any issues fixed, addressed or otherwise related to this pull request, including issue numbers or hard links for issues in other repos._
_You can delete these instructions once you have written your PR description._

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [ ] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

**Comment by gbggrant** [comment]
retest this please

File: src/main/java/picard/arrays/illumina/BpmToNormalizationManifestCsv.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    @Argument(shortName = StandardOptionDefinitions.OUTPUT_SHORT_NAME, doc = "The output (bpm.csv) file to write.")
+    public File OUTPUT;
+
+    @Override
+    protected int doWork() {
+        IOUtil.assertFileIsReadable(INPUT);
+        IOUtil.assertFileIsReadable(CLUSTER_FILE);
+        IOUtil.assertFileIsWritable(OUTPUT);
+
+        InfiniumEGTFile infiniumEGTFile;
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/BpmToNormalizationManifestCsv.java:62]: final
```suggestion
        final InfiniumEGTFile infiniumEGTFile;
```

File: src/main/java/picard/arrays/illumina/BpmToNormalizationManifestCsv.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        try {
+            infiniumEGTFile = new InfiniumEGTFile(CLUSTER_FILE);
+        } catch (IOException e) {
+            throw new PicardException("Error reading cluster file '" + CLUSTER_FILE.getAbsolutePath() + "'", e);
+        }
+
+        IlluminaBPMFile bpmFile;
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/BpmToNormalizationManifestCsv.java:69]:
```suggestion
        final IlluminaBPMFile bpmFile;
```

File: src/main/java/picard/arrays/illumina/BpmToNormalizationManifestCsv.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        try (final DataInputStream inputStream = new DataInputStream(new FileInputStream(INPUT))) {
+            bpmFile = new IlluminaBPMFile(inputStream);
+        } catch (IOException e) {
+            throw new PicardException("Error reading bpm file '" + INPUT.getAbsolutePath() + "'", e);
+        }
+
+        DecimalFormat df = new DecimalFormat(("0.0000"));
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/BpmToNormalizationManifestCsv.java:76]:
```suggestion
        final DecimalFormat df = new DecimalFormat(("0.0000"));
```

File: src/main/java/picard/arrays/illumina/IlluminaBPMFile.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    /**
+     * Main parsing method.
+     *
+     * @throws IOException thrown when there is a problem reading the Bead Pool Manifest
+     */
+    private void parse() throws IOException {
+
+        stream.mark(0);
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/IlluminaBPMFile.java:40]: Is this necessary?

File: src/main/java/picard/arrays/illumina/IlluminaBPMFile.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+            int version = parseInt();
+            final int versionFlag = 0x1000;
+            if ((version & versionFlag) == versionFlag) {
+                version ^= versionFlag;
+            }
+            if (version > 5 || version < 3) {
+                throw new PicardException("Unsupported BPM version (" + version + ")");
+            }
+            manifestName = parseString();
+            if (version > 1) {
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/IlluminaBPMFile.java:66]: true

File: src/main/java/picard/arrays/illumina/IlluminaBPMFile.java [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+        // Read the names
+        String[] names = new String[numLoci];
+        for (int i = 0; i < numLoci; i++) {
+            names[i] = parseString();
+        }
+
+        // Read the normalization ids.
+        allNormalizationIds = parseByteArrayAsInts(numLoci);
+        Set<Integer> uniqueNormalizationIdsSet = new TreeSet<>();
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/IlluminaBPMFile.java:89]: Could probably use a hashset here
```suggestion
        final Set<Integer> uniqueNormalizationIdsSet = new HashSet<>();
```

File: src/main/java/picard/arrays/illumina/IlluminaBPMFile.java [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+        // Initialize the locus entries
+        locusEntries = new IlluminaBPMLocusEntry[numLoci];
+        for (int i = 0; i < numLoci; i++) {
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/IlluminaBPMFile.java:93]: may not need to initialize as null

File: src/main/java/picard/arrays/illumina/IlluminaBPMFile.java [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+            locusEntries[i] = null;
+        }
+
+        // Read the locus entries.
+        for (int i = 0; i < numLoci; i++) {
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/IlluminaBPMFile.java:98]: Consider refactoring into a separate method

File: src/main/java/picard/arrays/illumina/IlluminaBPMLocusEntry.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
+package picard.arrays.illumina;
+
+/**
+ * A simple class to represent a locus entry in an Illumina Bead Pool Manifest (BPM) file
+ */
+public class IlluminaBPMLocusEntry {
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/IlluminaBPMLocusEntry.java:6]: Consider passing the InfiniumDataFile into the constructor and parsing there

File: src/main/java/picard/arrays/illumina/IlluminaBPMLocusEntry.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    String genomeBuild;
+    String source;
+    String sourceVersion;
+    String sourceStrand;
+    String sourceSeq;           // CSV files or BPM files with version 4 data block
+    String topGenomicSeq;       // CSV files or BPM files with version 4 data block
+
+    int expClusters;
+    int intensityOnly;
+    int assayType;              // Identifies type of assay (0 - Infinium II , 1 - Infinium I (A/T), 2 - Infinium I (G/C)
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/IlluminaBPMLocusEntry.java:31]:
```suggestion
   // Identifies type of assay (0 - Infinium II , 1 - Infinium I (A/T), 2 - Infinium I (G/C)
    int assayType;
```

File: src/main/java/picard/arrays/illumina/InfiniumEGTFile.java [hunk 11]
```diff
@@ -151,16 +157,11 @@ private void parseFloats(float[] array) throws IOException {
...
-        skipString();
-        // skip dataCreated
-        skipString();
+        gencallVersion = parseString();
+        clusterVersion = parseString();
+        callVersion = parseString();
+        normalizationVersion = parseString();
+        dateCreated = parseString();
         // skip isWGT
         skipBoolean();
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/InfiniumEGTFile.java:166]: parseBoolean() ?

File: src/main/java/picard/arrays/illumina/InfiniumGTCFile.java [hunk 12]
```diff
@@ -82,7 +84,10 @@
     public static final byte AB_CALL = 2;
     public static final byte BB_CALL = 3;
 
-    private final InfiniumNormalizationManifest normalizationManifest;
+    // Normalization Ids as pulled from the BPM file
+    private int[] allNormalizationIds = null;
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/InfiniumGTCFile.java:88]: null is default

File: src/main/java/picard/arrays/illumina/InfiniumGTCFile.java [hunk 13]
```diff
@@ -163,12 +169,17 @@ public void close() throws IOException {
...
 
-    InfiniumGTCFile(final DataInputStream gtcStream) throws IOException {
-        super(gtcStream, true);
-        this.normalizationManifest = null;
-        parse();
-    }
+    private void loadNormalizationIds(final File bpmFile) {
+        IlluminaBPMFile illuminaBPMFile;
+        try (final DataInputStream inputStream = new DataInputStream(new FileInputStream(bpmFile))) {
+            illuminaBPMFile = new IlluminaBPMFile(inputStream);
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/InfiniumGTCFile.java:175]:
```suggestion
            illuminaBPMFile = new IlluminaBPMFile(bpmFile);
```
And make IlluminaBPMFile autocloseable

File: src/main/java/picard/arrays/illumina/InfiniumGTCFile.java [hunk 14]
```diff
@@ -221,7 +232,7 @@ private void parse() throws IOException {
             stream.close();
         }
 
-        if ((normalizationManifest != null) && (normalizationManifest.getNormIds() != null)) {
+        if (allNormalizationIds != null) {
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/InfiniumGTCFile.java:235]: always true

File: src/main/java/picard/arrays/illumina/InfiniumGTCFile.java [hunk 15]
```diff
@@ -230,15 +241,14 @@ private void normalizeIntensities() {
...
-        final int[] normIds = normalizationManifest.getNormIds();
         for (int i = 0; i < rawXIntensities.length; i++) {
             final int rawX = rawXIntensities[i];
             final int rawY = rawYIntensities[i];
 
             final int normId;
             int normIndex = -1;
-            if ((normIds != null) && (normIds.length > i)) {
-                normId = normIds[i];
+            if ((allNormalizationIds != null) && (allNormalizationIds.length > i)) {
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/InfiniumGTCFile.java:250]: simplify
```suggestion
            if (allNormalizationIds.length > i) {
```

File: src/main/java/picard/arrays/illumina/InfiniumGTCFile.java [hunk 16]
```diff
@@ -495,7 +505,7 @@ private void parseTransformations(final InfiniumFileTOC toc) throws IOException
     }
 
     private void normalizeAndCalculateStatistics() {
-        if (normalizationManifest.getNormIds() != null) {
+        if (allNormalizationIds != null) {
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/InfiniumGTCFile.java:508]: always true

File: src/main/java/picard/arrays/illumina/BpmToNormalizationManifestCsv.java (hunk 2 above)
Comment (gbggrant) [reply src/main/java/picard/arrays/illumina/BpmToNormalizationManifestCsv.java:69]: done

File: src/main/java/picard/arrays/illumina/IlluminaBPMFile.java (hunk 5 above)
Comment (gbggrant) [reply src/main/java/picard/arrays/illumina/IlluminaBPMFile.java:66]: Thanks.

File: src/main/java/picard/arrays/illumina/IlluminaBPMFile.java (hunk 7 above)
Comment (gbggrant) [reply src/main/java/picard/arrays/illumina/IlluminaBPMFile.java:93]: that is true.

File: src/main/java/picard/arrays/illumina/InfiniumEGTFile.java (hunk 11 above)
Comment (gbggrant) [reply src/main/java/picard/arrays/illumina/InfiniumEGTFile.java:166]: yup

File: src/main/java/picard/arrays/illumina/InfiniumGTCFile.java (hunk 12 above)
Comment (gbggrant) [reply src/main/java/picard/arrays/illumina/InfiniumGTCFile.java:88]: Yup

File: src/main/java/picard/arrays/illumina/InfiniumGTCFile.java (hunk 13 above)
Comment (gbggrant) [reply src/main/java/picard/arrays/illumina/InfiniumGTCFile.java:175]: yup

File: src/main/java/picard/arrays/illumina/IlluminaBPMFile.java (hunk 4 above)
Comment (gbggrant) [reply src/main/java/picard/arrays/illumina/IlluminaBPMFile.java]: Nope - getting rid of it.

File: src/main/java/picard/arrays/illumina/IlluminaBPMFile.java (hunk 5 above)
Comment (gbggrant) [reply src/main/java/picard/arrays/illumina/IlluminaBPMFile.java]: Done

File: src/main/java/picard/arrays/illumina/IlluminaBPMFile.java (hunk 7 above)
Comment (gbggrant) [reply src/main/java/picard/arrays/illumina/IlluminaBPMFile.java]: Done

File: src/main/java/picard/arrays/illumina/InfiniumGTCFile.java (hunk 14 above)
Comment (gbggrant) [reply src/main/java/picard/arrays/illumina/InfiniumGTCFile.java]: Done

File: src/main/java/picard/arrays/illumina/InfiniumGTCFile.java (hunk 15 above)
Comment (gbggrant) [reply src/main/java/picard/arrays/illumina/InfiniumGTCFile.java]: Done.

File: src/main/java/picard/arrays/illumina/InfiniumGTCFile.java (hunk 16 above)
Comment (gbggrant) [reply src/main/java/picard/arrays/illumina/InfiniumGTCFile.java:506]: Done

File: src/main/java/picard/arrays/illumina/IlluminaBPMFile.java (hunk 8 above)
Comment (gbggrant) [reply src/main/java/picard/arrays/illumina/IlluminaBPMFile.java:101]: I've refactored it into a separate method.

File: src/main/java/picard/arrays/illumina/IlluminaBPMLocusEntry.java (hunk 9 above)
Comment (gbggrant) [reply src/main/java/picard/arrays/illumina/IlluminaBPMLocusEntry.java:6]: I'm going to hold off on that for now.  Maybe in another PR

**Comment by gbggrant** [comment]
@jessicaway @pshapiro4broad I think I've addressed all of your comments.  Can you please take another look?

File: src/main/java/picard/arrays/illumina/IlluminaBPMLocusEntry.java [hunk 17]
```diff
@@ -0,0 +1,80 @@
+package picard.arrays.illumina;
+
+/**
+ * A simple class to represent a locus entry in an Illumina Bead Pool Manifest (BPM) file
+ */
+public class IlluminaBPMLocusEntry {
+    String ilmnId;              // IlmnID (probe identifier) of locus
```
Comment (jessicaway) [inline src/main/java/picard/arrays/illumina/IlluminaBPMLocusEntry.java:7]: generally prefer comments on separate lines

Comment (gbggrant) [reply src/main/java/picard/arrays/illumina/IlluminaBPMLocusEntry.java:7]: Thanks.  Forgot to get them all

**Pull request merged**